### ISSUE TITLE
i18next languages isn't initialized when running component in unit test

### ISSUE THREAD

**Issue by StarpTech** [isaachinman/next-i18next#605]
## Describe the bug
Hello, I'm trying to mount my component in jest test suite. The component is really simple it just push a new route with the exported `next-i18next` router.

```jsx
const ComponentFilter = ({}) => {
    Router.push(
      {
        pathname: PathMap.components.pathname,
        query: {
          categorySlug: Router.query.categorySlug,
        }
      },
      PathMap.components.as([Router.query.categorySlug])
    );

return HERE YOUR INPUT FIELD;
}
```
test
```jsx
test('Click checkbox', async () => {
  const { getByLabelText } = render(<FilterComponent />);
  const checkbox = getByLabelText('components:compatibleCheckboxLabel');
  fireEvent.click(checkbox);
});
```

I got the error:
```
Render component list with active filter, toggle filter
lca:     TypeError: Cannot read property '0' of undefined
lca:       17 |     setMyrobot(myRobotCompatible);
lca:       18 | 
lca:     > 19 |     Router.push(
lca:          |            ^
lca:       20 |       {
lca:       21 |         pathname: PathMap.components.pathname,
lca:       22 |         query: {
lca:       at Object.Router.<computed> [as push] (../../node_modules/next-i18next/dist/commonjs/router/wrap-router.js:52:63)
```

this occurs because the property `i18n.languages` is undefined.

https://github.com/isaachinman/next-i18next/blob/785c8bcbe35c3e31bc8ae80342542e961e3b07ee/src/router/wrap-router.ts#L38

How should I do it in my tests?

## Occurs in next-i18next version
2.1.0

## Expected behaviour
The i18n instance should be initialized and ready when next-i18next is created.

Related:
https://github.com/i18next/i18next/blob/ca38b449529bb25232590b8054abf5e5026854e3/src/i18next.js#L249

**Comment by isaachinman** [comment]
I don't really understand the problem yet. Can you post a full code sample?

**Comment by StarpTech** [comment]
I found a workaround. We can set the **lng** property based on an environment variable before running the tests.

```
    "test:ui": "cross-env I18NEXT_LNG=en jest",
```
and
```
  lng: process.env.I18NEXT_LNG,
```

This will enforce i18next to initialize the properties. While it works it would be nice if next-i18next would provide some support out of the box. The docs around testing are also sparse. I will help.

**Comment by StarpTech** [comment]
@isaachinman just setup jest + next.js and try to push the router in a button event handler. The test will fail because `i18n.languages` is undefined. That happens because i18next isn't initialized before the router method are wrapped.

**Comment by StarpTech** [comment]
jest on the server not with puppeteer

**Comment by StarpTech** [comment]
@isaachinman That's the same reason why poeple see

```
react-i18next:: i18n.languages were undefined or empty undefined
```

when starting their next.js app.

Related: https://github.com/isaachinman/next-i18next/issues/562

**Comment by isaachinman** [comment]
> That happens because i18next isn't initialized before the router methods are wrapped

There's now a promise on the NextI18Next class called `initPromise` that can be awaited. Let me know how you want to proceed with this issue.

**Comment by StarpTech** [comment]
The issue can't be fixed on consumer side. Accessing `i18n.languages` before i18next has initialized all resources is wrong.

The fix is to await i18next.init before wrapping the router methods.

**Comment by StarpTech** [comment]
The reason to do it is that our language detectors are async not sync.

**Comment by isaachinman** [comment]
> The issue can't be fixed on consumer side

> The fix is to await i18next.init before wrapping the router methods.

How does this make sense, given that non-test functionality is _not_ broken, even without awaiting the init promise?

> The reason to do it is that our language detectors are async not sync

Can you explain a bit more why this is related?

**Comment by StarpTech** [comment]
> How does this make sense, given that non-test functionality is not broken, even without awaiting the init promise?

Because we don't have proper E2E tests. There isn't a single test to test this module as the user would consume it in unit tests. 

The `simple example` is run inside puppeteer not in the same process. That's why we never experience a bug.

**Comment by StarpTech** [comment]
> Can you explain a bit more why this is related?

Checkout the link in the issue description. i18next is not immediately initialized when using async detectors and `setLng` is deferred where `languages` is set.

**Comment by StarpTech** [comment]
I could fix it with following options:

```js
  // This is required to set a language when tests are running
  // otherwise next-i18next will throw an error that "i18n.languages" isn't set during wraping the router methods.
  // https://github.com/isaachinman/next-i18next/issues/605
  lng: process.env.I18NEXT_LNG,
  initImmediate: !process.env.I18NEXT_LNG,
```

**Comment by isaachinman** [comment]
@StarpTech There's now a promise on the `NextI18Next` class which can be awaited. How would you like to proceed here?

**Comment by eric-burel** [comment]
@StarpTech do you have a sample test?

Here is my i18n file:
```
import NextI18Next from "next-i18next";

const NextI18NextInstance = new NextI18Next({
  defaultLanguage: "en",
  otherLanguages: ["fr"],
  lng: process.env.I18NEXT_LNG,
  // ensure sync loading
  initImmediate: !process.env.I18NEXT_LNG
});
NextI18NextInstance.i18n.languages = ["en", "fr"];

export default NextI18NextInstance;

/* Optionally, export class methods as named exports */
export const { appWithTranslation, withTranslation } = NextI18NextInstance;
```
My unit test:
```
  test("translates content", () => {
    const ToTranslate = ({ t }) => <div>{t("welcome")}</div>;
    ToTranslate.getInitialProps = async () => ({
      namespacesRequired: ["common"]
    });
    const Translated = withTranslation("common")(ToTranslate);
    const { getByText, getByRole, container, asFragment } = render(
      <Translated />
    );
    expect(container.textContent).toMatch(`welcome!`);
  });
```

It does not translate but I probably did not setup the test properly.

**Comment by StarpTech** [comment]
Hi @eric-burel that's hard to evaluate. Did you set up mocks for the `react-i18next` library? What testing library do you use?

In your specific example you didn't wrap it with `appWithTranslation`.

**Comment by StarpTech** [comment]
> There's now a promise on the NextI18Next class which can be awaited. How would you like to proceed here?

@isaachinman it doesn't help. Please check it out and test it by yourself.

**Comment by isaachinman** [comment]
Hi there, I have just released `next-i18next@8.0.0`, which uses the internationalised routing built into NextJs v10+. This new version of `next-i18next` is quite different, thus making this PR no longer relevant. Thanks!
